src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala

```diff
@@ -18,43 +19,55 @@ import scala.util.control.NonFatal
 private[tracker] object TaskOpProcessorImpl {
 
   /**
-    * Maps a task status update to the appropriate [[TaskOpProcessor.Action]].
+    * Maps a [[TaskStateOp]] to the appropriate [[TaskStateChange]].
     *
     * @param directTaskTracker a TaskTracker instance that goes directly to the correct taskTracker
     *                          without going through the WhenLeaderActor indirection.
     */
-  class StatusUpdateActionResolver(clock: Clock, directTaskTracker: TaskTracker) {
+  class TaskStateOpResolver(directTaskTracker: TaskTracker) {
     private[this] val log = LoggerFactory.getLogger(getClass)
 
     /**
-      * Maps the UpdateStatus action to
+      * Maps the TaskStateOp
       *
-      * * a Action.Fail if the task does not exist OR ELSE
-      * * a Action.Noop if the task does not have to be changed OR ELSE
-      * * an Action.Expunge if the TaskStatus update indicates a terminated task OR ELSE
-      * * an Action.Update if the tasks existed and the TaskStatus contains new information OR ELSE
+      * * a TaskStateChange.Failure if the task does not exist OR ELSE
+      * * delegates the TaskStateOp to the existing task that will then determine the state change
       */
-    def resolve(taskId: Task.Id, status: TaskStatus)(
-      implicit ec: ExecutionContext): Future[Action] = {
+    def resolve(op: TaskStateOp)(implicit ec: ExecutionContext): Future[TaskStateChange] = {
+      // FIXME (3221): Create and Expunge are mostly to restore states.
+      op match {
+        case op: TaskStateOp.Create              => updateIfNotExists(op.taskId, op.task)
+        case op: TaskStateOp.ForceExpunge        => updateExistingTask(op)
```

Comment:
At least IDEA by default put annoying (yet sometimes helpful) warnings in there when I did that in the past, therefore I moved away from doing it completely.